src/components/geocoder-panel.js

```diff
@@ -0,0 +1,243 @@
+import Processors from 'processors';
+import {fitBounds, addDataToMap, removeDataset} from 'actions';
+
+const GEOCODER_DATASET_NAME = 'geocoder_dataset';
+const QUERY_PARAMS = {};
+const GEO_OFFSET = 0.1;
+const ICON_LAYER = {
+  id: 'geocoder_layer',
+  type: 'icon',
+  config: {
+    label: 'Geocoder Layer',
+    color: [255, 0, 0],
+    dataId: GEOCODER_DATASET_NAME,
+    columns: {
+      lat: 'lt',
+      lng: 'ln',
+      icon: 'icon',
+      label: 'text'
+    },
+    isVisible: true,
+    hidden: true,
+    visConfig: {
+      radius: 80
+    }
+  }
+};
+
+const GeocoderPanelContent = styled.div`
+  background-color: ${props => props.theme.panelBackground};
+  padding: 0.5em 1em;
+  position: absolute;
+  top: 1em;
+  left: 50%;
+  margin-left: -20em;
+  width: 40em;
+  z-index: 10000;
+  box-sizing: border-box;
+`;
+
+const GeocoderWrapper = styled.div`
+  color: ${props => props.theme.textColor};
+  font-size: ${props => props.theme.fontSize};
+
+  .react-geocoder {
+    position: relative;
+  }
+
+  .react-geocoder input {
+    ${props => props.theme.secondaryInput};
+  }
+
+  .react-geocoder-results {
+    background-color: ${props => props.theme.panelBackground};
+    position: absolute;
+    width: 43.5em;
+  }
+
+  .react-geocoder-item {
+    ${props => props.theme.dropdownListItem};
+    ${props => props.theme.textTruncate};
+  }
+
+  .remove-layer {
+    background: transparent;
+    border: none;
+    bottom: 28px;
+    color: ${props => props.theme.textColor};
+    cursor: pointer;
+    display: inline;
+    font-size: 16px;
+    padding: 2px 8px;
+    position: absolute;
+    right: 16px;
+
+    :hover,
+    :focus,
+    :active {
+      background: transparent !important;
+      border: none;
+      box-shadow: 0;
+      color: ${props => props.theme.textColor};
+      opacity: 0.6;
+      outline: none;
+    }
+  }
+`;
+
+function generateGeocoderDataset(lat, lon, text) {
+  return {
+    data: Processors.processRowObject([
+      {
+        lt: lat,
+        ln: lon,
+        icon: 'place',
+        text
+      }
+    ]),
+    id: GEOCODER_DATASET_NAME,
+    info: {
+      hidden: true,
+      id: GEOCODER_DATASET_NAME,
+      label: GEOCODER_DATASET_NAME
+    }
+  };
+}
+
+function isValid(key) {
+  return /pk\..*\..*/.test(key);
+}
+
+export class GeocoderPanel extends Component {
+  static propTypes = {
+    isGeocoderEnabled: PropTypes.bool.isRequired,
+    mapboxApiAccessToken: PropTypes.string.isRequired
+  };
+
+  constructor() {
+    super();
+    this.onSelected = this.onSelected.bind(this);
+    this.removeMarker = this.removeMarker.bind(this);
+
+    this.state = {
+      selectedGeoItem: null
+    };
+  }
+
+  removeGeocoderDataset() {
+    this.props.dispatch(removeDataset(GEOCODER_DATASET_NAME));
+  }
+
+  onSelected(viewport = null, geoItem) {
+    const {
+      center: [lon, lat],
+      text,
+      bbox
+    } = geoItem;
+    this.removeGeocoderDataset();
+    this.props.dispatch(
```

Comment:
this can be problematic, because you are adding dataset with the same id over and over again. I would check if the dataset is already exist (there is already some geocode marker), if it doesn't, call`addDataToMap`, with the dataset and layer.  if it does, create a new `visStateAction` call it `appendRowsToDataset`

make the updater as this:
```js
export function appendRowsToDataset(state, action) {
  const {rows, dataId} = action;
  if (!state.datasets[dataId]) {
    // dataset doesnot exist
    console.warn(`dataset.${dataId} doesn't exist`);
    return state;
  }
  const oldDataset = {
    ...state.datasets[dataId]
  };

  // create a temp dataset and run filter on it
  const tempDataset = {
    id: dataId,
    filterRecord: {},
    allData: rows,
    allIndexes: rows.map((_, i) => i)
  };
  const layersToFilter = (state.layers || []).filter(l => l.config.dataId === dataId);
  const appliedFilters = (state.filters || []).filter(d => shouldApplyFilter(d, dataId));
  
  const tempfiltered = filterDataset(tempDataset, appliedFilters, layersToFilter);
  const allData = oldDataset.allData.concat(rows);
  const offsetIdx = i => i + oldDataset.allData.length;

  const newDataset = {
    ...oldDataset,
    allData,
    allIndexes: allData.map((_, i) => i),
    filteredIndex: oldDataset.filteredIndex.concat(tempfiltered.filteredIndex.map(offsetIdx)),
    filteredIndexForDomain: oldDataset.filteredIndexForDomain.concat(tempfiltered.filteredIndex.map(offsetIdx))
  }
  const newState = {
    ...state,
    datasets:
    {
      ...state.datasets,
      [dataId]: newDataset
    }
  }

  // update layers associatred with this data
  return updateAllLayerDomainData(newState, [dataId]);
}
```

I just worte this one, haven't test it yet, but i think this function will also help enable data streaming in the future

Reply:
@heshan0131 is this related to mouse move bug? I'm still trying to figure it out. I can always ad _.uniqueId to dataset. Also I'm removing that dataset before adding. I haven't seen any problems yet.

Reply:
@heshan0131 every time new dataset is generated since we only need one value and one row. Please check https://deploy-preview-1068--keplergl.netlify.com/ to see how Geocoder search works now.